### PR TITLE
Convert variance sample to udaf

### LINKED ISSUE #10667: Convert `Variance Sample` to UDAF
### Is your feature request related to a problem or challenge?

Part of #8708 
1. Move variance aggregate expression in `datafusion/physical-expr/src/aggregate/variance.rs` to `functions-aggregate`.
2. Add expression api in `roundtrip_expr_api` in `datafusion/proto/tests/cases/roundtrip_logical_plan.rs`

Similar to #10372 #10418

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by yyin-dev**
## Which issue does this PR close?

Closes #10667 .

## Are these changes tested?

- [x] New proto roundtrip tests.
- [x] Moved tests into sqllogictests.

### CODE REVIEW

**Comment by yyin-dev** [comment]
@jayzhan211  I'm working on a change, but can you help me understand the semantics here:

```
# csv_query_distinct_variance
query R
SELECT var(distinct c2) FROM aggregate_test_100
----
2.5

statement error DataFusion error: This feature is not implemented: VAR\(DISTINCT\) aggregations are not available
SELECT var(c2), var(distinct c2) FROM aggregate_test_100
```
Why should the first query succeed but not the second one?

**Comment by jayzhan211** [comment]
> @jayzhan211 I'm working on a change, but can you help me understand the semantics here:
> 
> ```
> # csv_query_distinct_variance
> query R
> SELECT var(distinct c2) FROM aggregate_test_100
> ----
> 2.5
> 
> statement error DataFusion error: This feature is not implemented: VAR\(DISTINCT\) aggregations are not available
> SELECT var(c2), var(distinct c2) FROM aggregate_test_100
> ```
> 
> Why should the first query succeed but not the second one? Feel free to point me to any SQL / datafusion doc.

I think it is because of optimize rule `SingleDistinctToGroupBy`, this rule convert distinct to group by, so the first query is no longer `distinct`, you can try adding `explain` to see the optimized logical plan.

**Comment by yyin-dev** [comment]
> > @jayzhan211 I'm working on a change, but can you help me understand the semantics here:
> > ```
> > # csv_query_distinct_variance
> > query R
> > SELECT var(distinct c2) FROM aggregate_test_100
> > ----
> > 2.5
> > 
> > statement error DataFusion error: This feature is not implemented: VAR\(DISTINCT\) aggregations are not available
> > SELECT var(c2), var(distinct c2) FROM aggregate_test_100
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Why should the first query succeed but not the second one? Feel free to point me to any SQL / datafusion doc.
> 
> I think it is because of optimize rule `SingleDistinctToGroupBy`, this rule convert distinct to group by, so the first query is no longer `distinct`, you can try adding `explain` to see the optimized logical plan.

Ah that makes sense. Thanks!

**Comment by yyin-dev** [comment]
> > @jayzhan211 I'm working on a change, but can you help me understand the semantics here:
> > ```
> > # csv_query_distinct_variance
> > query R
> > SELECT var(distinct c2) FROM aggregate_test_100
> > ----
> > 2.5
> > 
> > statement error DataFusion error: This feature is not implemented: VAR\(DISTINCT\) aggregations are not available
> > SELECT var(c2), var(distinct c2) FROM aggregate_test_100
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Why should the first query succeed but not the second one? Feel free to point me to any SQL / datafusion doc.
> 
> I think it is because of optimize rule `SingleDistinctToGroupBy`, this rule convert distinct to group by, so the first query is no longer `distinct`, you can try adding `explain` to see the optimized logical plan.

I'm thinking about the right way to implement error-raising. Before migration, the logic was implemented in `physical-exp/src/aggregate/build_in.rs:create_aggregate_expr` as a match statement.

After migration, the error should probably be raised in `phyical-expr-common/src/aggregate/mod.rs:create_aggregate_expr`. There are two options: 

1. Get the udaf's name and implement similar logic. This is simpler but less principled?

2. Adds a `support_distinct` to the `AggregateUDFImpl` trait. This feels like a better solution.

What do you think?

**Comment by jayzhan211** [comment]
> > > @jayzhan211 I'm working on a change, but can you help me understand the semantics here:
> > > ```
> > > # csv_query_distinct_variance
> > > query R
> > > SELECT var(distinct c2) FROM aggregate_test_100
> > > ----
> > > 2.5
> > > 
> > > statement error DataFusion error: This feature is not implemented: VAR\(DISTINCT\) aggregations are not available
> > > SELECT var(c2), var(distinct c2) FROM aggregate_test_100
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Why should the first query succeed but not the second one? Feel free to point me to any SQL / datafusion doc.
> > 
> > 
> > I think it is because of optimize rule `SingleDistinctToGroupBy`, this rule convert distinct to group by, so the first query is no longer `distinct`, you can try adding `explain` to see the optimized logical plan.
> 
> I'm thinking about the right way to implement error-raising. Before migration, the logic was implemented in `physical-exp/src/aggregate/build_in.rs:create_aggregate_expr` as a match statement.
> 
> After migration, the error should probably be raised in `phyical-expr-common/src/aggregate/mod.rs:create_aggregate_expr`. There are two options:
> 
> 1. Get the udaf's name and implement similar logic. This is simpler but less principled?
> 2. Adds a `fun support_distinct(&self) -> bool` to the `AggregateUDFImpl` trait. This feels like a better solution.
> 
> What do you think?

We have `is_distinct` in `AccumulatorArgs`, so you can return errors if you found `is_distinct` is true

**Comment by jayzhan211** [comment]
@yyin-dev There are some error left to fix

You can try `./dev/rust_lint.sh`, `cargo test --test sqllogictests` and `cargo test --lib --tests --bins --features avro,json` to pass the CI

**Comment by yyin-dev** [comment]
> @yyin-dev There are some error left to fix
> 
> You can try `./dev/rust_lint.sh`, `cargo test --test sqllogictests` and `cargo test --lib --tests --bins --features avro,json` before running the CI

Thanks! I just ran the 3 commands and fixed all problems. Can you trigger the CI again? I don't think I can trigger it myself.

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs [hunk 1]
```diff
@@ -651,6 +653,7 @@ async fn roundtrip_expr_api() -> Result<()> {
         covar_pop(lit(1.5), lit(2.2)),
         sum(lit(1)),
         median(lit(2)),
+        var_sample(lit(2.2)),
```
Comment (alamb) [inline datafusion/proto/tests/cases/roundtrip_logical_plan.rs:656]: 💯

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs [hunk 2]
```diff
@@ -31,8 +31,10 @@ use datafusion::datasource::TableProvider;
 use datafusion::execution::context::SessionState;
 use datafusion::execution::runtime_env::{RuntimeConfig, RuntimeEnv};
 use datafusion::execution::FunctionRegistry;
-use datafusion::functions_aggregate::expr_fn::{covar_pop, covar_samp, first_value};
+use datafusion::functions_aggregate::covariance::{covar_pop, covar_samp};
+use datafusion::functions_aggregate::expr_fn::first_value;
```
Comment (alamb) [inline datafusion/proto/tests/cases/roundtrip_logical_plan.rs:35]: Can you please change this to `use` the exports from `datafusion::functions_aggregate::expr_fn` (which I think is the intended external facing API)?

File: datafusion/proto/src/logical_plan/from_proto.rs [hunk 3]
```diff
@@ -149,7 +149,6 @@ impl From<protobuf::AggregateFunction> for AggregateFunction {
             protobuf::AggregateFunction::Count => Self::Count,
             protobuf::AggregateFunction::ApproxDistinct => Self::ApproxDistinct,
             protobuf::AggregateFunction::ArrayAgg => Self::ArrayAgg,
-            protobuf::AggregateFunction::Variance => Self::Variance,
```
Comment (alamb) [inline datafusion/proto/src/logical_plan/from_proto.rs:152]: 🎉

**Review by alamb — APPROVED** [review]
Thank you @yyin-dev  and @jayzhan211  -- I think this looks really nice. I had one very minor comment about `use` but otherwise I think this is ready to go@

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs (hunk 2 above)
Comment (yyin-dev) [reply datafusion/proto/tests/cases/roundtrip_logical_plan.rs]: Updated to use exports from `expr_fn`.

**Review by jayzhan211 — APPROVED** [review]
👍

**Comment by jayzhan211** [comment]
Thanks @yyin-dev and @alamb

**Pull request merged**